ttnn/cpp/ttnn/tensor/types.hpp

```diff
@@ -407,206 +396,220 @@ struct BorrowedStorage {
-            buffers = other.buffers;
-            shapes = other.shapes;
-        }
+    MultiDeviceHostStorage(const MultiDeviceHostStorage &other) {
+        std::lock_guard<std::mutex> lock(mtx);
+        strategy = other.strategy;
+        buffers = other.buffers;
+        shapes = other.shapes;
+    }
 
-        MultiDeviceHostStorage &operator=(const MultiDeviceHostStorage &other) {
-            std::lock_guard<std::mutex> lock(mtx);
-            strategy = other.strategy;
-            buffers = other.buffers;
-            shapes = other.shapes;
-            return *this;
-        }
+    MultiDeviceHostStorage &operator=(const MultiDeviceHostStorage &other) {
+        std::lock_guard<std::mutex> lock(mtx);
+        strategy = other.strategy;
+        buffers = other.buffers;
+        shapes = other.shapes;
+        return *this;
+    }
 
-        MultiDeviceHostStorage &operator=( MultiDeviceHostStorage &&other) {
-            std::lock_guard<std::mutex> lock(mtx);
-            strategy = other.strategy;
-            buffers = other.buffers;
-            shapes = other.shapes;
-            return *this;
-        }
+    MultiDeviceHostStorage &operator=(MultiDeviceHostStorage &&other) {
+        std::lock_guard<std::mutex> lock(mtx);
+        strategy = other.strategy;
+        buffers = other.buffers;
+        shapes = other.shapes;
+        return *this;
+    }
 
-        bool operator == (const MultiDeviceHostStorage& other) {
-            return this->strategy == other.strategy and this->buffers == other.buffers and this->shapes == other.shapes;
-        }
+    bool operator==(const MultiDeviceHostStorage &other) {
+        return this->strategy == other.strategy and this->buffers == other.buffers and this->shapes == other.shapes;
+    }
 
     static constexpr auto attribute_names = std::forward_as_tuple();
     const auto attribute_values() const { return std::forward_as_tuple(); }
 
-        // Helper Functions - Getters and setters to get/modify storage attributes. These are needed to
-        // preinitialize empty tensor handles and use/populate them in the worker threads.
-        void insert_buffer_and_shape_for_device(int buffer_index, const OwnedBuffer buffer, const Shape shape) {
-            std::lock_guard<std::mutex> lock(mtx);
-            buffers[buffer_index] = buffer;
-            shapes[buffer_index] = shape;
-        }
-
-        OwnedBuffer get_buffer(int buffer_index) const {
-            std::lock_guard<std::mutex> lock(mtx);
-            TT_ASSERT(buffer_index < buffers.size(), "Buffer not found for buffer_index " + std::to_string(buffer_index));
-            return buffers[buffer_index];;
-        }
+    // Helper Functions - Getters and setters to get/modify storage attributes. These are needed to
+    // preinitialize empty tensor handles and use/populate them in the worker threads.
+    void insert_buffer_and_shape_for_device(int buffer_index, const OwnedBuffer buffer, const Shape shape) {
```

Comment:
const OwnerBuffer& buffer, const Shape& shape
We don't want to increae decrease atomic counters every call.